### ISSUE TITLE
[bug]: listunspent shows spent outputs

### ISSUE THREAD

**Issue by kallerosenbaum** [lightningnetwork/lnd#8786]
### Background

This might be a duplicate of https://github.com/lightningnetwork/lnd/issues/5959, but not sure so I'll open a new ticket.

We have two LND nodes of the same version 0.17.2 (but later they got upgraded to 0.17.5). We call them lnd-main and lnd-routing.

Both these nodes recently reported and still report outputs they don't have. I'm focusing on lnd-routing here (leaving out all correct outputs):

```
# date; lncli listunspent 1|grep outpoint                                                                                                                 
Tue May 28 10:01:04 UTC 2024                                                                                                                                             
                        "outpoint": "a4c248d5344ea4aa1ac9800e46aad117a77aa27b8e6501c1f84c6c52765e392e:0",
                        "outpoint": "685fdb1818fce6fe2d4fac960044e490ed6ab6965a45e454570e2424577fca82:0",
                        "outpoint": "4e8e54a38f11576cf9b0599c6eee2e2321532f40eea584aa410e16cfefd1dee9:0",
                        "outpoint": "9e83457fc8fe53b537330dcd32d99d62588d61fc3ea7b6b2ec392fd3c7b842b5:0",
                        "outpoint": "a592d2a1ef610e084d44e102fcb844a966cbe97557ec7b92a6b555c3ea294ac5:0",
                        "outpoint": "1254868d20ad31359144b684a483a64e44afb100bab3d7402897667228305d91:0",
                        "outpoint": "575faeb98596d4ff9064323924bf90b1672a36f1737405b70a800be908fc3fb1:0",
                        "outpoint": "fbb02dfcecf7c5ca7d2afe60d5b35b884151a58cf21dace1e63aab3217853ed1:0",
                        "outpoint": "68e7ecf111f1570078547d4d38de0034c806687e7e46391f87e08c3983e78e0d:0",
                        "outpoint": "6d138e3729973d1dd9078e6ff15cc10c40707be1a390c110a6ed0ea1cb01fec4:0",
```

These are already spent by tx `61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f`, which was broadcast 2024-04-09 and confirmed on 2024-05-15.

### Your environment

* LND 0.17.2 when issue appeared, 0.17.5 when we discovered it.,
* Linux lnd-main-0 6.2.0-1017-aws # 17~22.04.1-Ubuntu SMP Fri Nov 17 21:07:13 UTC 2023 x86_64 Linux
* bitcoin core 26 when issue appeared, 27 when we discovered the issue
* running in kubernetes

### Steps to reproduce

When searching through the logs I found this line right after a restart on 2024-05-21:

```
2024-05-21 12:10:16.516 [INF] LNWL: 61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f: broadcast failed because of: double spend: -25: bad-txns-inputs-missingorspent
2024-05-21 12:10:16.519 [INF] LNWL: Removed invalid transaction: (*wire.MsgTx)(0xc002faae10)({
Version: (int32) 2,
TxIn: ([]*wire.TxIn) (len=10 cap=10) {
(*wire.TxIn)(0xc003992000)({
PreviousOutPoint: (wire.OutPoint) 1254868d20ad31359144b684a483a64e44afb100bab3d7402897667228305d91:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  a3 fb b2 d9 44 03 50 9b  ab 3c cd 88 b5 07 c9 e6  |....D.P..<......|
00000010  b2 bf 64 2d 46 f5 7f f9  cb 82 6f f0 7b bb 4f 63  |..d-F.....o.{.Oc|
00000020  b1 fd cd d5 28 33 0c 87  5a ea 0d d0 cd b7 4d 93  |....(3..Z.....M.|
00000030  e5 22 27 bd da ab 15 1f  54 d7 3b 50 9b b2 22 e8  |."'.....T.;P..".|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc003992060)({
PreviousOutPoint: (wire.OutPoint) 4e8e54a38f11576cf9b0599c6eee2e2321532f40eea584aa410e16cfefd1dee9:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  0f d1 41 fe 8e 7a c2 19  ec 20 95 62 37 6f 0a d4  |..A..z... .b7o..|
00000010  53 00 c4 cd 18 b8 06 64  31 5c 48 96 bb 68 30 5f  |S......d1\H..h0_|
00000020  12 e1 ae 95 7c 2e 13 3c  b2 e3 87 20 80 05 87 ea  |....|..<... ....|
00000030  b6 76 d6 46 61 bd 64 f4  01 7c 90 53 fc 53 1a 9f  |.v.Fa.d..|.S.S..|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc0039920c0)({
PreviousOutPoint: (wire.OutPoint) 575faeb98596d4ff9064323924bf90b1672a36f1737405b70a800be908fc3fb1:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  48 67 77 34 27 b2 a0 e2  f3 93 0d a9 44 1a 6b 67  |Hgw4'.......D.kg|
00000010  2a dc 04 a3 e5 2b da 47  43 a9 b4 b5 5b 6d c1 08  |*....+.GC...[m..|
00000020  b5 21 7e 3e d9 dd 63 e7  30 a9 65 d3 d4 4f 76 ea  |.!~>..c.0.e..Ov.|
00000030  af 2b d7 b8 26 bc ac ff  03 d7 bf 44 9f 9c 34 b8  |.+..&......D..4.|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc003992120)({
PreviousOutPoint: (wire.OutPoint) 685fdb1818fce6fe2d4fac960044e490ed6ab6965a45e454570e2424577fca82:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  6a 90 cb ea ba 6a 57 55  e6 e9 45 fb c0 16 d2 00  |j....jWU..E.....|
00000010  4a 6a f1 23 f1 93 f2 b4  e4 3d b9 b5 7f 17 a8 5d  |Jj.#.....=.....]|
00000020  66 ff 45 d2 59 49 12 04  48 a0 0e cb cb 73 52 b5  |f.E.YI..H....sR.|
00000030  75 c7 9b 08 97 c5 03 51  0d 1a ef 19 89 08 05 c1  |u......Q........|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc003992180)({
PreviousOutPoint: (wire.OutPoint) 68e7ecf111f1570078547d4d38de0034c806687e7e46391f87e08c3983e78e0d:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  5e df 1a cf 05 1d e4 12  5b 63 5a 62 c2 95 c8 72  |^.......[cZb...r|
00000010  e9 25 a1 c7 64 85 d9 9c  30 81 59 cb 64 00 55 dc  |.%..d...0.Y.d.U.|
00000020  99 da 6f b5 d4 b7 6b 3d  7a 1e df 37 56 12 fd 7c  |..o...k=z..7V..||
00000030  a1 da d0 0b 65 4b cd 88  77 f0 cc ea f3 30 a9 c0  |....eK..w....0..|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc0039921e0)({
PreviousOutPoint: (wire.OutPoint) 6d138e3729973d1dd9078e6ff15cc10c40707be1a390c110a6ed0ea1cb01fec4:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  29 3e d1 77 60 29 21 22  21 33 34 0a 74 f6 9e 5f  |)>.w`)!"!34.t.._|
00000010  2d 1d 6b 36 12 8a 98 ba  2a 10 da 30 09 85 c2 4e  |-.k6....*..0...N|
00000020  e0 d0 c9 0a 70 f7 5d 1c  ac 93 b3 c4 83 a8 cc a3  |....p.].........|
00000030  75 59 74 4e 34 b9 eb 88  b4 ce ca 65 af 48 3c d2  |uYtN4......e.H<.|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc003992240)({
PreviousOutPoint: (wire.OutPoint) 9e83457fc8fe53b537330dcd32d99d62588d61fc3ea7b6b2ec392fd3c7b842b5:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  20 f1 4e 7d ac ee 8d 3e  59 05 8c a6 fc fe 9e 08  | .N}...>Y.......|
00000010  b2 54 99 7d c9 5b 38 20  c9 d0 33 b9 3c f0 a4 71  |.T.}.[8 ..3.<..q|
00000020  89 3d 28 5a de 9b 7c b6  b2 c9 a7 86 e3 19 f2 7d  |.=(Z..|........}|
00000030  bf f1 9c bd 1c bf 3d 37  48 a5 a3 a3 35 60 7b 81  |......=7H...5`{.|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc0039922a0)({
PreviousOutPoint: (wire.OutPoint) a4c248d5344ea4aa1ac9800e46aad117a77aa27b8e6501c1f84c6c52765e392e:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  64 58 99 13 57 72 36 20  9c f6 ca c0 93 6e 73 30  |dX..Wr6 .....ns0|
00000010  d3 3f 7c e3 32 b7 f1 dc  04 c3 40 cc 15 e0 5c 8e  |.?|.2.....@...\.|
00000020  6f b0 18 a1 64 85 4a bb  b7 ec 75 99 3f 1d 7b 8f  |o...d.J...u.?.{.|
00000030  f0 61 81 b0 31 ef 3c b3  71 bd 31 11 e7 68 78 56  |.a..1.<.q.1..hxV|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc003992300)({
PreviousOutPoint: (wire.OutPoint) a592d2a1ef610e084d44e102fcb844a966cbe97557ec7b92a6b555c3ea294ac5:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  16 8a 55 46 39 54 2a eb  22 61 99 cf 2c 6b 12 49  |..UF9T*."a..,k.I|
00000010  c0 95 91 eb 52 e0 34 5c  47 d7 20 35 43 7d 73 eb  |....R.4\G. 5C}s.|
00000020  71 55 41 e9 2a 1b a6 3f  65 d5 97 69 f7 88 5d 2c  |qUA.*..?e..i..],|
00000030  d6 50 c3 64 2b fb 96 95  6e 45 f2 f5 8d f8 87 98  |.P.d+...nE......|
}
},
Sequence: (uint32) 4294967293
}),
(*wire.TxIn)(0xc003992360)({
PreviousOutPoint: (wire.OutPoint) fbb02dfcecf7c5ca7d2afe60d5b35b884151a58cf21dace1e63aab3217853ed1:0,
SignatureScript: ([]uint8) {
},
Witness: (wire.TxWitness) (len=1 cap=1) {
([]uint8) (len=64 cap=64) {
00000000  4b ed 3e 24 fe b2 0e 12  b5 ab 8b a9 73 a2 33 36  |K.>$........s.36|
00000010  b2 43 a2 de 55 a8 11 b1  21 77 a6 58 e6 20 db 1f  |.C..U...!w.X. ..|
00000020  be fa 27 a4 b0 15 0e af  29 bc fd b3 8c 94 ce 0a  |..'.....).......|
00000030  70 c5 c0 d6 21 f9 11 f2  8c 65 cd ac 47 15 1b 14  |p...!....e..G...|
}
},
Sequence: (uint32) 4294967293
})
},
TxOut: ([]*wire.TxOut) (len=1 cap=1) {
(*wire.TxOut)(0xc0013e8b80)({
Value: (int64) 407085,
PkScript: ([]uint8) (len=22 cap=22) {
00000000  00 14 43 29 c6 53 f3 8c  b9 ed 83 28 6c 7e cb 12  |..C).S.....(l~..|
00000010  cf 73 45 ff 58 36                                 |.sE.X6|
}
})
},
LockTime: (uint32) 0
})

2024-05-21 12:10:16.520 [INF] LNWL: Removed invalid transaction: 0200000000010a915d30287266972840d7b3ba00b1af444ea683a484b644913531ad208d8654120000000000fdffffffe9ded1efcf160e41aa84a5ee402f5321232eee6e9c59b0f96c57118fa3548e4e0000000000fdffffffb13ffc08e90b800ab7057473f1362a67b190bf2439326490ffd49685b9ae5f570000000000fdffffff82ca7f5724240e5754e4455a96b66aed90e4440096ac4f2dfee6fc1818db5f680000000000fdffffff0d8ee783398ce0871f39467e7e6806c83400de384d7d54780057f111f1ece7680000000000fdffffffc4fe01cba10eeda610c190a3e17b70400cc15cf16f8e07d91d3d9729378e136d0000000000fdffffffb542b8c7d32f39ecb2b6a73efc618d58629dd932cd0d3337b553fec87f45839e0000000000fdffffff2e395e76526c4cf8c101658e7ba27aa717d1aa460e80c91aaaa44e34d548c2a40000000000fdffffffc54a29eac355b5a6927bec5775e9cb66a944b8fc02e1444d080e61efa1d292a50000000000fdffffffd13e851732ab3ae6e1ac1df28ca55141885bb3d560fe2a7dcac5f7ecfc2db0fb0000000000fdffffff012d360600000000001600144329c653f38cb9ed83286c7ecb12cf7345ff58360140a3fbb2d94403509bab3ccd88b507c9e6b2bf642d46f57ff9cb826ff07bbb4f63b1fdcdd528330c875aea0dd0cdb74d93e52227bddaab151f54d73b509bb222e801400fd141fe8e7ac219ec209562376f0ad45300c4cd18b80664315c4896bb68305f12e1ae957c2e133cb2e38720800587eab676d64661bd64f4017c9053fc531a9f01404867773427b2a0e2f3930da9441a6b672adc04a3e52bda4743a9b4b55b6dc108b5217e3ed9dd63e730a965d3d44f76eaaf2bd7b826bcacff03d7bf449f9c34b801406a90cbeaba6a5755e6e945fbc016d2004a6af123f193f2b4e43db9b57f17a85d66ff45d25949120448a00ecbcb7352b575c79b0897c503510d1aef19890805c101405edf1acf051de4125b635a62c295c872e925a1c76485d99c308159cb640055dc99da6fb5d4b76b3d7a1edf375612fd7ca1dad00b654bcd8877f0cceaf330a9c00140293ed177602921222133340a74f69e5f2d1d6b36128a98ba2a10da300985c24ee0d0c90a70f75d1cac93b3c483a8cca37559744e34b9eb88b4ceca65af483cd2014020f14e7dacee8d3e59058ca6fcfe9e08b254997dc95b3820c9d033b93cf0a471893d285ade9b7cb6b2c9a786e319f27dbff19cbd1cbf3d3748a5a3a335607b81014064589913577236209cf6cac0936e7330d33f7ce332b7f1dc04c340cc15e05c8e6fb018a164854abbb7ec75993f1d7b8ff06181b031ef3cb371bd3111e76878560140168a554639542aeb226199cf2c6b1249c09591eb52e0345c47d72035437d73eb715541e92a1ba63f65d59769f7885d2cd650c3642bfb96956e45f2f58df8879801404bed3e24feb20e12b5ab8ba973a23336b243a2de55a811b12177a658e620db1fbefa27a4b0150eaf29bcfdb38c94ce0a70c5c0d621f911f28c65cdac47151b1400000000
```

This transaction is `61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f`.

The same happened on lnd-main too when it restarted about 30 minutes prior to lnd-routing, but I won't list the details about that issue here. I'll provide whatever you need upon request.

Before we discovered this we upgraded both nodes to 0.17.5 on May 23.

I'm hesitant to restart with --reset-wallet-transactions because I don't know the consequences. The wallet is very old and I'm not sure how long the rescanning would take and what limitations apply on the node while doing it.

**Comment by ViktorTigerstrom** [comment]
Hi @kallerosenbaum,

Thanks for the issue report! I've been trying to investigate the issue to see if I can spot the root cause.

I have some initial questions and requests:
1. The log line(s) you've included, i.e.:
`2024-05-21 12:10:16.516 [INF] LNWL: 61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f: broadcast failed because of: double spend: -25: bad-txns-inputs-missingorspent`
Does this log line reoccur for every restart of `lnd`, or was this showed only once?
2. Could you possibly include a few more of the log lines that are before and after the log lines you've included? I.e. the log lines right before `2024-05-21 12:10:16.516` and after `2024-05-21 12:10:16.520`.

Finally, you're correct that restarting lnd with the `--reset-wallet-transactions` flag set will most likely solve your issue. Just like you say though, the restart will take a long time given the age of the node and and will result in limited functionality during the rescan. Unfortunately it's hard to give any real estimates of how long the rescan will take as that depends on many factors.

**Comment by kallerosenbaum** [comment]
@ViktorTigerstrom thanks! Correction: we didn't upgrade from 0.17.2 to 0.17.5, we upgraded from 0.17.3 to 0.17.5.

1. It seems to be only this one restart. I searched the logs for `'61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f: broadcast failed' || '0.17.'` and this came out:
```
2024-05-21 11:42:08.185 [INF] LTND: Version: 0.17.3-beta commit=v0.17.3-beta, build=production, logging=default, debuglevel=info
2024-05-21 12:10:16.516 [INF] LNWL: 61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f: broadcast failed because of: double spend: -25: bad-txns-inputs-missingorspent
2024-05-23 12:21:10.200 [INF] LTND: Version: 0.17.5-beta commit=v0.17.5-beta, build=production, logging=default, debuglevel=info
2024-05-23 14:30:08.538 [INF] LTND: Version: 0.17.5-beta commit=v0.17.5-beta, build=production, logging=default, debuglevel=info
```
2. See attached file
[lnd-routing-2024-05-21-noonish.log](https://github.com/lightningnetwork/lnd/files/15495246/lnd-routing-2024-05-21-noonish.log)

**Comment by kallerosenbaum** [comment]
@ViktorTigerstrom I'm probably going to restart with `--reset-wallet-transactions` early next week, so if you want anything from me before that, please shout. Maybe useful troubleshooting data will get lost when we reset?

**Comment by ViktorTigerstrom** [comment]
Apologies for the late reply @kallerosenbaum! Thank you for the extended logs, we should hopefully have what we need for the further investigation with that. You can go ahead with the restart with `--reset-wallet-transactions` to resolve the issue on your side. Thanks!

**Comment by kallerosenbaum** [comment]
I should probably mention that we sweep transactions in a maybe somewhat unusual way.

```
        // 1. Create a template psbt with the exact set of inputs and an output of 1000 sats (small non-dust number),
        //    and 1 sat_vbyte fee.
        // 2. Call fundPsbt, whilch will fill in required signing info in the psbt inputs, but not sign it. It also
        //    potentially adds a change output.
        // 3. Call finalizePsbt to sign the psbt. Now we have a fully signed psbt that lets us figure out the vSize of
        //    that transaction.
        // 4. Then we need to copy the template psbt and delete any change output LND added and set the
        //    single output's amount to <sumofinputs> - <vsizeofnewtransactions> * feeRate
        // 5. Sign
        // 6. Make sanity checks, etc., and publish
```

**Comment by kallerosenbaum** [comment]
@ViktorTigerstrom ran --reset-wallet-transactions and it's back to normal. But I have a feeling the issue will resurface, since it happened on both our nodes. I think it might have to do with our onchain sweeping method (described in the comment above) which runs on both nodes.  The method should be fine, but maybe it isn't handled well by LND for some reason. Or am I doing something that I really shouldn't be doing?

**Comment by yyforyongyu** [comment]
Just curious, how long did it take to restart lnd with the `--reset-wallet-transactions` flag set?

> But I have a feeling the issue will resurface, since it happened on both our nodes.

I think there's some investigation work needs to be done, think we should target this for 0.19.0? cc @saubyk 


> // 1. Create a template psbt with the exact set of inputs and an output of 1000 sats (small non-dust number),
> //    and 1 sat_vbyte fee.

As for the sweeping method, I don't see anything wrong here - is this used for UTXO aggregation purpose? How did you get this set of inputs?

**Comment by saubyk** [comment]
> I think there's some investigation work needs to be done, think we should target this for 0.19.0? cc @saubyk

Tagged for 0.19

**Comment by kallerosenbaum** [comment]
@yyforyongyu 
> how long did it take to restart lnd with the --reset-wallet-transactions flag set?

```
2024-06-11 07:40:27.005 [INF] LNWL: Started rescan from block 0000000000000000000dc34778ca1059404b0761795e5a91e3ae84c16d855b30 (height 537669) for 25764 addresses
2024-06-11 07:40:30.480 [INF] LNWL: Rescanned through block 000000000000000000200b39910d1e9566501a63026d4a88588aa6f7f0c4eb8b (height 537678)
2024-06-11 07:45:24.161 [INF] LNWL: Rescanned through block 00000000000000000011b3e92e82e0f9939093dccc3614647686c20e5ebe3aa6 (height 540000)
... and so on ...
2024-06-12 01:43:12.018 [INF] LNWL: Finished rescan for 25764 addresses (synced to block 0000000000000000000247841cf08aff2f584491f7738e5a41df8188a72d9901, height 847548)
```

> is this used for UTXO aggregation purpose?

Yes, we aggregate to an external wallet.

> How did you get this set of inputs?

`listunspent 1` over grpc. Then we remove outputs that cost more than a threshold percentage to spend and make sure to keep reserved funds in LND.

I think it might have to do with the relatively low fees we set on the transactions. The one described in this issue (`61f8f4b41196b01a3acb3eee16a81f890540d3b624ed2b0efa730520947f272f`) took more than 5 weeks to confirm (when it was CPFP'd).

**Comment by yyforyongyu** [comment]
1 sat/vb is indeed pretty low - during a low-fee environment it should be enough to get the tx in the mempool, then it can be CPFPed. During a high-fee environment, however, peers may not relay the tx (you can check peers' minimal relay feerate via `bitcoind-cli getpeerinfo | jq '.[].minfeefilter'`).

There's the `lncli sendcoins` that may be more suitable in this case, as it performs a series of checks,
- validate min relay feerate is met;
- validate the constructed tx passes mempool constraints (`testmempoolaccept`);
- actively rebroadcast the tx in case it falls out of the mempool.

Meanwhile we will look into the `listunspent` issue in 0.19!

**Comment by kallerosenbaum** [comment]
> 1 sat/vb is indeed pretty low 

We don't set 1 sat/vB. We set fee according to bitcoin core's `estimatesmartfee 1008`. The 1 sat/vB is just for the initial "template transaction", the actual fee we want. See step 4 in the algo above. The tx in this issue had 5.16 sat/vB.

> There's the lncli sendcoins that may be more suitable in this case

I opted not to use that since it doesn't give me enough control over the UTXOs to spend.

**Comment by kallerosenbaum** [comment]
FYI: We've been hit by this issue again on one of our nodes. I've kept the full logs for future investigation.

**Comment by kallerosenbaum** [comment]
@yyforyongyu I've been able to reproduce this on regtest. I use a docker-compose with bitcoin core 27.0 and lnd 0.17.5 to get the environment up, then I repeat the following until the bug triggers:

1. Add two UTXOs to the LND wallet
2. Sweep the UTXOs according to algorithm [described here](https://github.com/lightningnetwork/lnd/issues/8786#issuecomment-2162211451)
3. Stop the LND node
4. mine the sweep transaction with 6 confirmations
5. Start the LND node

It can take up to 40 iterations before it happens. It varies. The sweep in step 2 fails with "output already spent".

I'm sure this can be further simplified, for example use only 1 UTXO or reduce the number of confirmations, but at least this is a start.

I run LND with the following conf:
```
        --listen=0.0.0.0:9735
        --externalip=lnd-1
        --rpclisten=0.0.0.0:10009
        --bitcoin.active
        --bitcoin.node=bitcoind
        --bitcoin.regtest
        --bitcoind.rpcuser=test
        --bitcoind.rpcpass=password
        --bitcoind.rpchost=bitcoin:18443
        --bitcoind.zmqpubrawblock=tcp://bitcoin:18501
        --bitcoind.zmqpubrawtx=tcp://bitcoin:18502
        --norest
        --protocol.wumbo-channels
```

**Comment by kallerosenbaum** [comment]
@yyforyongyu When this issue gets fixed, is there a chance that the fix can, in addition to prevent future issues, also recover from ongoing issues? I mean, is it possible to heal from existing false UTXOs, for example by going through all UTXOs during upstart and check against bitcoin core if they are valid and if not, remove them?

The reason for this is that a rescan took ~20h last time (https://github.com/lightningnetwork/lnd/issues/8786#issuecomment-2171748397).

We currently have this issue on both our nodes and it causes lots of headache, especially with accounting.

**Comment by yyforyongyu** [comment]
@kallerosenbaum Def - I'm investigating this issue this week and will come up with a fix to handle both existing and future cases.

**Comment by yyforyongyu** [comment]
This should be fixed by #9609 now, feel free to reopen if it occurs again.

**Comment by kallerosenbaum** [comment]
@yyforyongyu Thanks. I will try to reproduce this issue on `0.19.0-beta.rc1`, but it's going to take a day or so because of the probabilitic nature of my test. In the meantime, I'll just note here that existing false UTXOs, that were created by `0.18.5-beta` aren't cleared by upgrading to `0.19.0-beta.rc1`. You said something about making this issue self-heal simply by upgrading: https://github.com/lightningnetwork/lnd/issues/8786#issuecomment-2531277119

**Comment by kallerosenbaum** [comment]
@yyforyongyu I was able to trigger this bug on 0.19.0-beta.rc1 (the docker image), so I'd like to reopen the issue, but I have no permission to do so. Here's some evidence:
```
$ docker exec -it docker-lnd-0-1 lncli -network=regtest listunspent                                                                                                         
{                                                                                                                                                                                                                                               
    "utxos": [                                                                                                                                                                                                                                  
        {                                                                                                                                                                                                                                       
            "address_type": 4,                                                                                                                                                                                                                  
            "address": "bcrt1ptwan8mm5fvyn6js0tydmv98m696d404cmasvajwktar5e3us380qmstqag",                                                                                                                                                      
            "amount_sat": 10018,                                                                                                                                                                                                                
            "pk_script": "51205bbb33ef744b093d4a0f591bb614fbd174dabeb8df60cec9d65f474cc79089de",                                                                                                                                                
            "outpoint": "9db7466781bc5e95654c557c0efe23a225049e477e0bc08fa99259dfade6840e:1",                                                                                                                                                   
            "confirmations": 1                                                                                                                                                                                                                  
        },                                                                                                                                                                                                                                      
        {                                                                                                                                                                                                                                       
            "address_type": 4,                                                                                                                                                                                                                  
            "address": "bcrt1pl2x0a7yqsju7vkx9jvcar7xrhtjtu04p5ts2vl76qdx0j7ka3gasj5nf6k",                                                                                                                                                      
            "amount_sat": 10015,                                                                                                                                                                                                                
            "pk_script": "5120fa8cfef88084b9e658c59331d1f8c3bae4be3ea1a2e0a67fda034cf97add8a3b",                                                                                                                                                
            "outpoint": "0dd748dce119ee39b005f78dd6bba62d3dde8f0d808a423b09868b4053c7b1ec:0",                                                                                                                                                   
            "confirmations": 22                                                                                                                                                                                                                 
        },                                                                                                                                                                                                                                      
        {                                                                                                                                                                                                                                       
            "address_type": 4,                                                                                                                                                                                                                  
            "address": "bcrt1pqfvm84yam7u7xc2pewczq6x3yzds64ekl73nydxng6adwknuvq4qyqepxj",                                                                                                                                                      
            "amount_sat": 10014,                                                                                                                                                                                                                
            "pk_script": "51200259b3d49ddfb9e36141cbb02068d1209b0d5736ffa33234d346bad75a7c602a",                                                                                                                                                
            "outpoint": "3082d58d48f7a23ce5385f546540b59f4b9010db723ebb653c90bddccfb73212:1",                                                                                                                                                   
            "confirmations": 29                                                                                                                                                                                                                 
        },                                                                                   
        ... and plenty more
```
And on the bitcoin node:
```
bitcoin@e87e12758ece:~$ bitcoin-cli -chain=regtest gettxout 9db7466781bc5e95654c557c0efe23a225049e477e0bc08fa99259dfade6840e 1
{
  "bestblock": "0e087875003c3a4b017776a88a3924bc59bf0fcca407e78f4f8b9d8b9d7f0f16",
  "confirmations": 1,
  "value": 0.00010018,
  "scriptPubKey": {
    "asm": "1 5bbb33ef744b093d4a0f591bb614fbd174dabeb8df60cec9d65f474cc79089de",
    "desc": "rawtr(5bbb33ef744b093d4a0f591bb614fbd174dabeb8df60cec9d65f474cc79089de)#2w7z7g2k",
    "hex": "51205bbb33ef744b093d4a0f591bb614fbd174dabeb8df60cec9d65f474cc79089de",
    "address": "bcrt1ptwan8mm5fvyn6js0tydmv98m696d404cmasvajwktar5e3us380qmstqag",
    "type": "witness_v1_taproot"
  },
  "coinbase": false
}
bitcoin@e87e12758ece:~$ bitcoin-cli -chain=regtest gettxout 0dd748dce119ee39b005f78dd6bba62d3dde8f0d808a423b09868b4053c7b1ec 0
bitcoin@e87e12758ece:~$ bitcoin-cli -chain=regtest gettxout 3082d58d48f7a23ce5385f546540b59f4b9010db723ebb653c90bddccfb73212 1
```

Version is 0.19.0-beta.rc1:
```
$ docker exec -it docker-lnd-0-1 lncli -network=regtest getinfo | grep vers
    "version": "0.19.0-beta.rc1 commit=v0.19.0-beta.rc1",
```

**Comment by yyforyongyu** [comment]
@kallerosenbaum Thanks for the feedback! Re-open this issue as it needs more investigation, we def will fix it in our incoming walletdb sqlization.

**Comment by kallerosenbaum** [comment]
I'd be happy to assist in any troubleshooting, since I'm eager to have this fixed. I can, for example, spend some time to simplify the test I wrote (described [above](https://github.com/lightningnetwork/lnd/issues/8786#issuecomment-2252708784)) to trigger this issue.

**Comment by kallerosenbaum** [comment]
I managed to distill the issue to this typescript test. There's lots of support code around this of course, but I think the code can serve as pseudo-code for you to reproduce in go or whatever you want. The difference from my previous tests is that here I don't use my home-brewed sweeper, but simply use `sendCoins` to send all funds. The test relies on a docker-compose, with a LND 0.19.0-beta.rc1 and Bitcoin Core 28.1.
```
        /**
         * This test is used to trigger https://github.com/lightningnetwork/lnd/issues/8786
         * It sometimes has to run up to 100 iterations before it fails.
         */
        test('Trigger bug 8786', async () => {
            let i = 0n;
            while (true) {
                console.log(`PREPARE UTXOS iteration ${i}`);
                // Create a UTXO, build a block, and await one confirmation
                await it.addUtxosAndConfirm(node0, [['TR', 10000n + i++]]);
                console.log('SWEEPING');
                // node0 is a thin layer on top of the grpc clients.
                await node0.sendCoins({
                    spend_unconfirmed: true,
                    sat_per_vbyte: 1,
                    addr: it.minerAddr,
                    send_all: true,
                });
                console.log('STOPPING');
                // calls stopDaemon. Docker-compose will restart this node when it dies.
                await node0.stop();
                console.log('STOPPED');
                await sleep(2000);
                console.log('WAITED 2s');
                // Build some blocks to confirm the sweep tx.
                await it.regTest.buildBlocks(6, it.minerAddr);
                console.log('ACTIVATE');
                // This will unlock the wallet and await state 'SERVER_ACTIVE'
                await node0.activate();
                console.log('ACTIVATED');
            }
        // Set the timeout for the test really high:
        }, 100000000);
```

**Comment by kallerosenbaum** [comment]
Here's my docker-compose.yml
```
services:
  bitcoin:
    image: bitrefill/bitcoin:28.1
    environment:
      BITCOIN_CONF: |
        chain=regtest
        listen=1
        fallbackfee=0.00001000
        maxtxfee=20000000
        txindex=1
        whitelist=noban@0.0.0.0/0
        rpcuser=test
        rpcpassword=password
        rpcallowip=0.0.0.0/0
        zmqpubrawblock=tcp://0.0.0.0:18501
        zmqpubrawtx=tcp://0.0.0.0:18502
        [regtest]
        rpcbind=0.0.0.0
    ports:
      - 48443:18443 # RPC port
      - 48444:18444 # P2P port

  lnd-0:
    build: ./lnd
    depends_on:
      bitcoin:
        condition: service_started
    ports:
      - 10009:10009
    environment:
      LND_ARGS: >
        --listen=0.0.0.0:9735
        --externalip=lnd-0
        --rpclisten=0.0.0.0:10009
        --bitcoin.active
        --bitcoin.node=bitcoind
        --bitcoin.regtest
        --bitcoind.rpcuser=test
        --bitcoind.rpcpass=password
        --bitcoind.rpchost=bitcoin:18443
        --bitcoind.zmqpubrawblock=tcp://bitcoin:18501
        --bitcoind.zmqpubrawtx=tcp://bitcoin:18502
        --norest
        --protocol.wumbo-channels
    restart: always
```
Note that the lnd image is built from lightninglabs/lnd image with this Dockerfile that creates a .lnd with static macaroon and certificate:
```
$ cat Dockerfile 
FROM lightninglabs/lnd:v0.19.0-beta.rc1

COPY ./dotlnd /root/.lnd

ENTRYPOINT lnd $LND_ARGS
```
and the structure
```
$ tree
.
├── Dockerfile
└── dotlnd
    ├── data
    │   └── chain
    │       └── bitcoin
    │           └── regtest
    │               ├── admin.macaroon
    │               └── macaroons.db
    ├── tls.cert
    └── tls.key
```

**Comment by kallerosenbaum** [comment]
It's interesting to see that utxos that were spent very long ago reappear seemingly out of nowhere. My latest version of the test is as follows:
```
        /**
         * This test is used to trigger https://github.com/lightningnetwork/lnd/issues/8786
         * It sometimes has to run up to 100 iterations before it fails.
         */
        test('Trigger bug 8786', async () => {
            let i = 0n;
            while (true) {
                console.log(`PREPARE UTXOS iteration ${i}`);
                // Create a UTXO, build a block, and await one confirmation
                await it.addUtxosAndConfirm(node0, [['TR', 10000n + i++]]);
                console.log('SWEEPING');
                // node0 is a thin layer on top of the grpc clients.
                await node0.sendCoins({
                    spend_unconfirmed: true,
                    sat_per_vbyte: 1,
                    addr: it.minerAddr,
                    send_all: true,
                });
                console.log('STOPPING');
                // calls stopDaemon. Docker-compose will restart this node when it dies.
                await node0.stop();
                console.log('STOPPED');
                // Build some blocks to confirm the sweep tx.
                await it.regTest.buildBlocks(1, it.minerAddr);
                await sleep(2000);
                console.log('ACTIVATE');
                // This will unlock the wallet and await state 'SERVER_ACTIVE'
                await node0.activate();
                console.log('ACTIVATED');
                // Verify that we have 0 UTXOs. Sooner or later this check will fail.
                const unspent = await node0.listUnspent({ min_confs: 1 });
                strict.equal(unspent!.utxos.length, 0);
            }
        // Set the timeout for the test really high:
        }, 100000000);
```
When I saw `strict.equal(unspent!.utxos.length, 0);` fail (with length 1 instead of the expected 0) I started manually probing `lncli -network=regtest listunspent` and this showed up:
```
{
    "utxos": [
        {
            "address_type": 4,
            "address": "bcrt1pnqgejer5csp4l06v53tvwhgv3xzj5fkxcndads5zjqq23eartevsy2naf7",
            "amount_sat": 10000,
            "pk_script": "51209811996474c4035fbf4ca456c75d0c89852a26c6c4dbd6c2829000a8e7a35e59",
            "outpoint": "1fe15f6968c5bc58c3c19543798f9088c8298656400443d1fe6fce602472460a:0",
            "confirmations": 102
        }
    ]
}
```
Note that the amount is 10000 (indicating that this is from iteration 0). The confirmations `102` also show that this is a really old output. It hasn't been shown by `listunspent` for  ~50 iterations, and then suddenly it shows up. What's more iteresting is, as I continue probing, more and more spent UTXOs start showing up in the exact order they were createad.
```
{
    "utxos": [
        {
            "address_type": 4,
            "address": "bcrt1pagkrgl940wcrw7zn647tlw5desrqg6dcva6399ulvxa7dvluamcsh4a0ur",
            "amount_sat": 10004,
            "pk_script": "5120ea2c347cb57bb0377853d57cbfba8dcc060469b8677512979f61bbe6b3fceef1",
            "outpoint": "958abd7f6de78a3489c43cd94434dbeca1a71e05c31b003e203d2fda6b93c4e0:1",
            "confirmations": 94
        },
        {
            "address_type": 4,
            "address": "bcrt1pagw55vc9nmjxxaneth6txc88j06r8t8h8jxveet5n8zss56g2nds2gyu0p",
            "amount_sat": 10003,
            "pk_script": "5120ea1d4a33059ee46376795df4b360e793f433acf73c8ccce57499c508534854db",
            "outpoint": "4f6a7366883c604c4252b6d1b603bc9de34dc327dbc60dd1eff0a63f105ce52d:0",
            "confirmations": 96
        },
        {
            "address_type": 4,
            "address": "bcrt1pxdc7gxned7478rhkml6c5wraynfrc8cnl42gtgwkj54e6s5pgd8qzy3ec7",
            "amount_sat": 10002,
            "pk_script": "51203371e41a796fabe38ef6dff58a387d24d23c1f13fd5485a1d6952b9d4281434e",
            "outpoint": "4c83a342dd49fcf192d742e6c15d9cda4ebee335d158b701c46e93e2aa51787a:0",
            "confirmations": 98
        },
        {
            "address_type": 4,
            "address": "bcrt1p6ls4lx0hscz4prx9hefepdprkm3f486ck3tm6yzk43fzjw9a5gyqf97hym",
            "amount_sat": 10001,
            "pk_script": "5120d7e15f99f78605508cc5be5390b423b6e29a9f58b457bd1056ac522938bda208",
            "outpoint": "5e2e75ffca783b95ee367493a128b9a9a1cf9028a3b18f23b8b66744c98b26ab:1",
            "confirmations": 100
        },
        {
            "address_type": 4,
            "address": "bcrt1pnqgejer5csp4l06v53tvwhgv3xzj5fkxcndads5zjqq23eartevsy2naf7",
            "amount_sat": 10000,
            "pk_script": "51209811996474c4035fbf4ca456c75d0c89852a26c6c4dbd6c2829000a8e7a35e59",
            "outpoint": "1fe15f6968c5bc58c3c19543798f9088c8298656400443d1fe6fce602472460a:0",
            "confirmations": 102
        }
    ]
}
```
And this keeps on happening (1 "UTXO" every few seconds) until all of them have reappeared, all the way up to:
```
{
    "utxos": [
        {
            "address_type": 4,
            "address": "bcrt1pfnvel4gsegfqg2ahqu0688yrphjseqxjlwumw6rerprr0auck6hq2lhu2d",
            "amount_sat": 10050,
            "pk_script": "51204cd99fd510ca12042bb7071fa39c830de50c80d2fbb9b76879184637f798b6ae",
            "outpoint": "d4c2d35a6b9a3e0e14340946f4557f134fcf9dedcf3469a33d15901283034a51:1",
            "confirmations": 2
        },
        {
            "address_type": 4,
            "address": "bcrt1ppr4e5zwpatu09tt4e5hf425e6sug4y6rgcrputcdkxf002gfd4kq9f728m",
            "amount_sat": 10049,
            "pk_script": "512008eb9a09c1eaf8f2ad75cd2e9aaa99d4388a934346061e2f0db192f7a9096d6c",
            "outpoint": "f3f734fc8182000b69e75892db3bbecf4812944913d47a9e5397ac38219b2024:0",
            "confirmations": 4
        },
        {
            "address_type": 4,
            "address": "bcrt1p2faztuljqajrlhf64pt933ggd5m9ezmv5lzxsfdxpyc56qsd7vwsuu4htr",
            "amount_sat": 10048,
            "pk_script": "5120527a25f3f207643fdd3aa85658c5086d365c8b6ca7c46825a609314d020df31d",
            "outpoint": "0fe0fe9969c4a61b140ddaf30e7eea1b084597d4b74c9e089036d9cc4f6c0908:0",
            "confirmations": 6
        },
        {
            "address_type": 4,
            "address": "bcrt1pwvnndrs6zumrytvxtxna6ztqcf5j3x94wfea2uzwcjn05ya2qtxsgtzr6y",
            "amount_sat": 10047,
            "pk_script": "51207327368e1a1736322d8659a7dd0960c2692898b57273d5704ec4a6fa13aa02cd",
            "outpoint": "b354b1cd8f10e89e3a80226d6edba05250524b41473560567e9045d19b0d250d:1",
            "confirmations": 8
        },
... and so on
```
During this time, nothing is printed on the lnd log. And of course, none of these outputs are real. Bitcoin Core denies their existence (using the oldest output in this example): 
```
$ docker exec -it docker-bitcoin-1 bitcoin-cli -chain=regtest gettxout 1fe15f6968c5bc58c3c19543798f9088c8298656400443d1fe6fce602472460a 0
<nothing printed>

# Just to show that bitcoin core is aware of the transaction
$ docker exec -it docker-bitcoin-1 bitcoin-cli -chain=regtest getrawtransaction 1fe15f6968c5bc58c3c19543798f9088c8298656400443d1fe6fce602472460a
0200000000010198c1e90de73af8b782d24cabd3c92343e95747bb0b495d484c3f1663339eef8a0000000000fdffffff0210270000000000002251209811996474c4035fbf4ca456c75d0c89852a26c6c4dbd6c2829000a8e7a35e594bca052a01000000225120b97e73edc16eccf41c77a090cbcccca5bfdd9d55a0f6ddfc570aa7d0df4c246202473044022069ee838aae4f599916b066f027f78fdc93d0edb0aed20825d26de8ab0a998bbe02201a582cf05720d2c428c4b378b504f5d04f73f5e106fe13c2557071f24fcd04eb012102882110a5f3824d09c4d244ebbaa40f18582528021cd2e38610167563e7895bd566000000
```

**Comment by yyforyongyu** [comment]
@kallerosenbaum Thanks for the detailed tests - I've successfully reproduced it, basically it happens with taproot outputs with 113 restarts, will look into how to fix it.

**Comment by kallerosenbaum** [comment]
@yyforyongyu In my case it varies between only a few restarts up to 60 restarts before it happens. I've never had to go as far as 113 restarts.

**Comment by kallerosenbaum** [comment]
@yyforyongyu do you currently see any way out of this without restarting with `--reset-wallet-transactions` as that will take around 20h to complete. Can I make lnd aware of the transactions that spent these UTXOs?

**Comment by yyforyongyu** [comment]
> [@yyforyongyu](https://github.com/yyforyongyu) do you currently see any way out of this without restarting with `--reset-wallet-transactions` as that will take around 20h to complete. Can I make lnd aware of the transactions that spent these UTXOs?

Unfortunately no, tho using `--reset-wallet-transactions` may be an overkill if you know the block of the first spending tx, in which case you can use `--recovery_window` in `lncli unlock --recovery_window xxx` to speed up.

**Comment by kallerosenbaum** [comment]
> tho using --reset-wallet-transactions may be an overkill if you know the block of the first spending tx, in which case you can use --recovery_window in lncli unlock --recovery_window xxx to speed up.

This seems wrong. The doc says:
```
   --recovery_window value  address lookahead to resume recovery rescan, value should be non-zero --  To recover all funds, this should be greater than the maximum number of consecutive, unused addresses ever generated by the wallet. (default: 0)
```
Which I read as the classic address "gap limit" that's common in any wallet that rescans the blockchain.

**Comment by yyforyongyu** [comment]
> Which I read as the classic address "gap limit" that's common in any wallet that rescans the blockchain.

Yeah you're right, was thinking of something else...how many addresses are there and what's the birthday block? Are you using kvdb or SQL? 20 hours is painfully too long.

**Comment by kallerosenbaum** [comment]
@yyforyongyu Last time I did a rescan, I documented it here: https://github.com/lightningnetwork/lnd/issues/8786#issuecomment-2171748397 I expect something similar this time.

We're on kvdb

**Comment by kallerosenbaum** [comment]
I restarted with `--reset-wallet-transactions` on one of our nodes. It took 15h. Last time was 18h, and I think it was the same node because both logs start from the same block:
```
2025-04-07 14:29:39.157 [INF] LNWL: Started rescan from block 0000000000000000000dc34778ca1059404b0761795e5a91e3ae84c16d855b30 (height 537669) for 26195 addresses
2025-04-07 14:29:43.471 [INF] LNWL: Rescanned through block 000000000000000000200b39910d1e9566501a63026d4a88588aa6f7f0c4eb8b (height 537678)
...
2025-04-08 05:29:30.605 [INF] LNWL: Rescanned through block 00000000000000000001f49fb5fb37753c06d78f1a811d707aebf6194bb147e5 (height 890000)
2025-04-08 05:33:42.097 [INF] LNWL: Finished rescan for 26195 addresses (synced to block 0000000000000000000112f2b27e3ffb4952e0857d275cd55cfbc1e4e4365721, height 891445)
```

**Comment by yyforyongyu** [comment]
> I restarted with --reset-wallet-transactions on one of our nodes. It took 15h. Last time was 18h, and I think it was the same node because both logs start from the same block:

Do you still find weird results from `listunspent` this time?

**Comment by kallerosenbaum** [comment]
> Do you still find weird results from listunspent this time?
No not yet, but last time I rescanned, new spent outputs showed up after a while. So I expect that to happen again soonish.

**Comment by yyforyongyu** [comment]
This is now fixed in #9693. Thanks again for the test case!

> No not yet, but last time I rescanned, new spent outputs showed up after a while. So I expect that to happen again soonish.

I expect it not to happen unless there' a restart, which has a fix in the upcoming release.

**Comment by kallerosenbaum** [comment]
@yyforyongyu Great! I tested this today on

* 0.18.5 to confirm that I can still trigger it, after 52 iterations
* on image lightninglabs/lnd:daily-testing-20250417 which doesn't include #9693. Issue triggered after 50 iterations
* on image lightninglabs/lnd:daily-testing-20250418, which has been running for 1000 iterations without failure

I strongly agree this is fixed. Thanks a lot! 🚀 

A few questions:

* I'm actually not 100% sure our issues in production are because of this bug. Now that you've fixed this issue, do you know if there's a way to check that this is the issue we're actually facing in prod? Some hints in the log or anything?
* Could you please try and give a rough explainer for what caused this issue? It seems super-weird.

**Comment by yyforyongyu** [comment]
@kallerosenbaum Glad it's fixed! As for the details, it's quite complicated, could you [DM](https://x.com/yyforyongyu) me to establish another channel so I can share them? I'm on slack/keybase/signal/telegram
